### PR TITLE
MTA: Clean up config, Enchanting Chamber update

### PR DESCRIPTION
**Pull request by Matthew-nop**
Changed config room names to reflect those on the wiki (e.g. Telekinetic Room -> Telekinetic Theatre). Changed config to use collapsible sections to reflect separation of different room options. Added a toggle for marking dragonstones in the enchantment room. Changed the indicator for dragonstone spawns from a hint arrow to a highlighted tile to allow highlighting of all dragonstones instead of nearest.

### CODE REVIEW

File: runelite-client/src/main/java/net/runelite/client/plugins/mta/MTAConfig.java [hunk 1]
```diff
@@ -81,18 +88,57 @@ default boolean alchemy()
...
+	}
+
+	@ConfigItem(
+			keyName = "defaultColor",
+			name = "Highlight color",
+			description = "The highlight color for dragonstones.",
+			position = 3,
+			section = enchantmentGroup
+	)
+	default Color dragonstoneColor()
```
Comment (Nightfirecat) [inline runelite-client/src/main/java/net/runelite/client/plugins/mta/MTAConfig.java:132]: This doesn't need to be configurable.

File: runelite-client/src/main/java/net/runelite/client/plugins/mta/MTAConfig.java [hunk 2]
```diff
@@ -81,18 +88,57 @@ default boolean alchemy()
...
 	}
 
+	@ConfigItem(
+			keyName = "highlightDragonstones",
+			name = "Highlight dragonstones",
+			description = "Enable markers indicating spawned dragonstones.",
+			position = 1,
+			section = enchantmentGroup
+	)
+	default boolean highlightDragonstones()
```
Comment (Nightfirecat) [inline runelite-client/src/main/java/net/runelite/client/plugins/mta/MTAConfig.java:108]: Add a "No marking" marking mode and this config is no longer needed.

File: runelite-client/src/main/java/net/runelite/client/plugins/mta/MTAConfig.java [hunk 3]
```diff
@@ -81,18 +88,57 @@ default boolean alchemy()
...
+		return Color.YELLOW;
+	}
+
+
 	@ConfigItem(
 		keyName = "graveyardEnable",
 		name = "Enable",
-		description = "Configures whether or not the graveyard room overlay is enabled.",
+		description = "Configures whether or not the Creature Graveyard overlay is enabled.",
+		position = 0,
```
Comment (Nightfirecat) [inline runelite-client/src/main/java/net/runelite/client/plugins/mta/MTAConfig.java:142]: Items which are alone in their config sections do not need positions assigned to them.

File: runelite-client/src/main/java/net/runelite/client/plugins/mta/enchantment/EnchantmentRoom.java [hunk 4]
```diff
@@ -67,67 +70,47 @@ public void onGameStateChanged(GameStateChanged gameStateChanged)
...
 	{
-		final TileItem item = itemDespawned.getItem();
-		final Tile tile = itemDespawned.getTile();
-
-		if (item.getId() == ItemID.DRAGONSTONE_6903)
-		{
-			WorldPoint location = tile.getWorldLocation();
-			log.debug("Removed dragonstone at {}", location);
-			dragonstones.remove(location);
+		if (config.enchantment()) {
```
Comment (Nightfirecat) [inline runelite-client/src/main/java/net/runelite/client/plugins/mta/enchantment/EnchantmentRoom.java]: Why is this added?

File: runelite-client/src/main/java/net/runelite/client/plugins/mta/enchantment/EnchantmentRoom.java [hunk 5]
```diff
@@ -34,22 +36,23 @@
...
+import net.runelite.client.ui.overlay.OverlayUtil;
 
 @Slf4j
 public class EnchantmentRoom extends MTARoom
 {
 	private static final int MTA_ENCHANT_REGION = 13462;
 
 	private final Client client;
-	private final List<WorldPoint> dragonstones = new ArrayList<>();
+	private final List<Tile> dragonstones = new ArrayList<>();
```
Comment (Nightfirecat) [inline runelite-client/src/main/java/net/runelite/client/plugins/mta/enchantment/EnchantmentRoom.java:55]: Why are tiles used here instead of WorldPoints like before? We only ever call `getLocalLocation()` of dragonstones.

File: runelite-client/src/main/java/net/runelite/client/plugins/mta/MTAConfig.java [hunk 6]
```diff
@@ -27,50 +27,86 @@
...
 	{
 		return true;
 	}
 
 	@ConfigItem(
-		keyName = "graveyard",
-		name = "Enable graveyard room",
-		description = "Configures whether or not the graveyard room overlay is enabled.",
-		position = 1
+			keyName = "enchantmentEnable",
```
Comment (Nightfirecat) [inline runelite-client/src/main/java/net/runelite/client/plugins/mta/MTAConfig.java]: Double check your formatter--these extra tabs are unwanted. `@ConfigItem(` should cause only one extra level of indentation.

**Review by Nightfirecat — CHANGES_REQUESTED** [review]

File: runelite-client/src/main/java/net/runelite/client/plugins/mta/MTAConfig.java (hunk 1 above)
Comment (Matthew-nop) [reply runelite-client/src/main/java/net/runelite/client/plugins/mta/MTAConfig.java:132]: Removed with latest commit.

File: runelite-client/src/main/java/net/runelite/client/plugins/mta/MTAConfig.java (hunk 2 above)
Comment (Matthew-nop) [reply runelite-client/src/main/java/net/runelite/client/plugins/mta/MTAConfig.java:108]: Good point, implemented in MTAConfig.

File: runelite-client/src/main/java/net/runelite/client/plugins/mta/MTAConfig.java (hunk 3 above)
Comment (Matthew-nop) [reply runelite-client/src/main/java/net/runelite/client/plugins/mta/MTAConfig.java:142]: Removed with latest commit.

File: runelite-client/src/main/java/net/runelite/client/plugins/mta/enchantment/EnchantmentRoom.java (hunk 4 above)
Comment (Matthew-nop) [reply runelite-client/src/main/java/net/runelite/client/plugins/mta/enchantment/EnchantmentRoom.java]: I added this as a guard since the enchantment room was the only room without them. I modified it to be more consistent with alchemy room (early return).

File: runelite-client/src/main/java/net/runelite/client/plugins/mta/enchantment/EnchantmentRoom.java (hunk 5 above)
Comment (Matthew-nop) [reply runelite-client/src/main/java/net/runelite/client/plugins/mta/enchantment/EnchantmentRoom.java:55]: I originally did this when testing, but forgot to revert it before pushing the changed. Reverted it back to WorldPoints in the latest commit.

File: runelite-client/src/main/java/net/runelite/client/plugins/mta/MTAConfig.java [hunk 7]
```diff
@@ -27,50 +27,105 @@
 import net.runelite.client.config.Config;
 import net.runelite.client.config.ConfigGroup;
 import net.runelite.client.config.ConfigItem;
+import net.runelite.client.config.ConfigSection;
 
-@ConfigGroup("mta")
+@ConfigGroup(MTAConfig.GROUP)
 public interface MTAConfig extends Config
 {
+	String GROUP = "mta";
```
Comment (Nightfirecat) [inline runelite-client/src/main/java/net/runelite/client/plugins/mta/MTAConfig.java:35]: Since this is never referenced elsewhere, this can be reverted.

**Review by Nightfirecat — CHANGES_REQUESTED** [review]
Additionally, can these changes be rebased into clean commits? One for code cleanup and another for the behavior changes. The diff is rather messy right now since there's fixup and merge commits littering the history.

File: runelite-client/src/main/java/net/runelite/client/plugins/mta/MTAConfig.java [hunk 8]
```diff
@@ -27,50 +27,105 @@
...
+		closedByDefault = true
+	)
+	String telekineticGroup = "telekinetic";
+
 	@ConfigItem(
-		keyName = "alchemy",
-		name = "Enable alchemy room",
-		description = "Configures whether or not the alchemy room overlay is enabled.",
-		position = 0
+		keyName = "alchemyEnable",
```
[the remark this answers is not in the document]
Comment (Matthew-nop) [reply runelite-client/src/main/java/net/runelite/client/plugins/mta/MTAConfig.java]: My bad, I didn't know that was the case. All keyName changes have been reverted.

File: runelite-client/src/main/java/net/runelite/client/plugins/mta/MTAConfig.java (hunk 7 above)
Comment (Matthew-nop) [reply runelite-client/src/main/java/net/runelite/client/plugins/mta/MTAConfig.java]: Reverted.

**Comment by Matthew-nop** [comment]
Done and done, hopefully this has organized the changes.

**Comment by Nightfirecat** [comment]
I've rewritten some of this code. (prioritize `inside()` checks over config checks, less code duplication, less config order shuffling, split into one more commit) I think this should be good to go now.

**Comment by Matthew-nop** [comment]
Thanks for going the extra mile. After reading what Adam said on the subject I'll make sure I'm more careful with config checks.

File: runelite-client/src/main/java/net/runelite/client/plugins/mta/enchantment/EnchantmentRoom.java [hunk 9]
```diff
@@ -106,12 +105,15 @@ private WorldPoint findNearestStone()
 	@Subscribe
 	public void onItemSpawned(ItemSpawned itemSpawned)
 	{
-		final TileItem item = itemSpawned.getItem();
-		final Tile tile = itemSpawned.getTile();
+		if (!inside() || !config.enchantment())
```
Comment (Adam-) [inline runelite-client/src/main/java/net/runelite/client/plugins/mta/enchantment/EnchantmentRoom.java:96]: Generally the way we implement on/off options for features is by having the on/off toggle the overlay while having the code which does the actual tracking run regardless. This is so it allows you to toggle the option on/off and have that worked as expected instead of requiring a relog. This is breaking that.

Also I don't really see any need to modify the existing code.

File: runelite-client/src/main/java/net/runelite/client/plugins/mta/enchantment/EnchantmentRoom.java [hunk 10]
```diff
@@ -143,4 +131,48 @@ public boolean inside()
 		return player != null && player.getWorldLocation().getRegionID() == MTA_ENCHANT_REGION
 			&& player.getWorldLocation().getPlane() == 0;
 	}
+
+	@Override
+	public void under(Graphics2D graphics)
+	{
+		final MarkingMode mode = config.dragonstoneMode();
```
Comment (Adam-) [inline runelite-client/src/main/java/net/runelite/client/plugins/mta/enchantment/EnchantmentRoom.java:138]: Prefer config checks to be after `inside()` as it is much more expensive.

File: runelite-client/src/main/java/net/runelite/client/plugins/mta/enchantment/EnchantmentRoom.java (hunk 9 above)
Comment (Matthew-nop) [reply runelite-client/src/main/java/net/runelite/client/plugins/mta/enchantment/EnchantmentRoom.java]: I've removed the config checks from here and `onItemDespawned()`. Although if there is a method to get all nearby items it may be better to repopulate the list of dragonstones when there's a config change.

File: runelite-client/src/main/java/net/runelite/client/plugins/mta/enchantment/EnchantmentRoom.java (hunk 10 above)
Comment (Matthew-nop) [reply runelite-client/src/main/java/net/runelite/client/plugins/mta/enchantment/EnchantmentRoom.java]: Changed with the latest commit

Comment (Matthew-nop) [reply runelite-client/src/main/java/net/runelite/client/plugins/mta/enchantment/EnchantmentRoom.java]: Changed with the latest commit

File: runelite-client/src/main/java/net/runelite/client/plugins/mta/enchantment/EnchantmentRoom.java (hunk 9 above)
Comment (Adam-) [reply runelite-client/src/main/java/net/runelite/client/plugins/mta/enchantment/EnchantmentRoom.java]: I would recommend just dropping this commit since it doesn't actually do anything and imo doesn't clean up anything either.

Comment (Matthew-nop) [reply runelite-client/src/main/java/net/runelite/client/plugins/mta/enchantment/EnchantmentRoom.java]: Sure, the changes are pretty inconsequential.

**Comment by Adam-** [comment]
Additionally I have removed the `inside()` check  in the enchantment room since it is unnecessary due to the MTASceneOverlay already checking that prior to caling `under`.

I've noticed this removes the hint arrow for the dragonstone, which probably a lot of users like, since it shows on the minimap and will flash.  This kind of seems like a step down?

**Comment by Matthew-nop** [comment]
Would it be preferable if there was another value in the enum which would maintain current behavior? I thought it seemed a bit out of place so I didn't include it, but if it's desired I can.

**Comment by Matthew-nop** [comment]
I've added another mode for the hint arrow functionality. Should this be made the default again?

**Comment by Nightfirecat** [comment]
I think having it add tile markers for all dragonstones and a hint arrow for the nearest one would be fine

**Comment by Matthew-nop** [comment]
> I think having it add tile markers for all dragonstones and a hint arrow for the nearest one would be fine


> I've noticed this removes the hint arrow for the dragonstone, which probably a lot of users like, since it shows on the minimap and will flash. This kind of seems like a step down?

I've changed it such that now the `ALL` mode will both highlight all dragonstones and assign a hint arrow to the nearest. Hopefully this can alleviate any concerns.
